ogr/ogrsf_frmts/hana/ogrhanadatasource.cpp

```diff
@@ -153,92 +152,100 @@ class OpenOptionsConstants
 
 CPLString BuildConnectionString(char** openOptions)
 {
+    // See notes for constructing connection string for HANA
+    // https://help.sap.com/docs/SAP_HANA_CLIENT/f1b440ded6144a54ada97ff95dac7adf/7cab593774474f2f8db335710b2f5c50.html
+
     std::vector<CPLString> params;
-    auto addParameter = [&](const char* optionName, const char* paramName) {
-        const char* paramValue =
-            CSLFetchNameValueDef(openOptions, optionName, nullptr);
+    bool isValid = true;
+    const CPLString specialChars("[]{}(),;?*=!@");
+
+    auto getOptValue = [&](const char* optionName, bool mandatory = false)
+    {
+        const char* paramValue = CSLFetchNameValueDef(openOptions, optionName, nullptr);
+        if (mandatory && paramValue == nullptr)
+        {
+            isValid = false;
+            CPLError(
+                CE_Failure, CPLE_AppDefined,
+                "Mandatory connection parameter '%s' is missing.", optionName);
+        }
+        return paramValue;
+    };
+
+    auto addParameter = [&](const char* paramName, const char* paramValue)
+    {
         if (paramValue == nullptr)
             return;
-        params.push_back(CPLString(paramName) + "=" + CPLString(paramValue));
+
+        CPLString value(paramValue);
+        bool hasSpecialChar = false;
+        for (char ch : value)
```

Comment:
maybe use https://cplusplus.com/reference/string/string/find_first_of/ ?

Reply:
Good point @rouault, thank you. I've pushed a new version.